Added Kallisto WDL

ambrosejcarr opened the pull request:
This PR adds: 

1. Kallisto task WDL
2. Kallisto mkref workflow
3. Kallisto testing notebook
4. Kallisto testing WDL

The main deliverable here is the task WDL, the rest are accessories involved in validating it's functionality.

jishuxu commented on pipelines/tasks/Kallisto.wdl:
```diff
@@ -0,0 +1,155 @@
+      --output-dir . \
+      --bootstrap-samples 100 \
+      --single \
+      --fragment-length 250 \
```
what's the fragment  estimation from? from 10X report or just arbitrary number?

ambrosejcarr replied:
arbitrary; updated now to be 10x numbers, as variables. 👍

jishuxu commented on pipelines/tasks/Kallisto.wdl:
```diff
@@ -0,0 +1,155 @@
+  Int disks = 10 + ceil((size(r1, "G") + size(r2, "G")) * 4 + size(index, "G") * 2)
+
+  command {
+    kallisto quant \
```
Can you add '--Pseudobam'  in cmd and output pseudoalignment bam file. cmd is something like 
kallisto quant -i index -o out --pseudobam r1.fastq r2.fastq | samtools view -Sb - > out.bam

ambrosejcarr replied:
Finding alllll them undocumented requirements!

```
[quant] fragment length distribution will be estimated from the data
Error: pseudobam is not compatible with running on many threads.
```

jishuxu replied:
they should do better job on their documentation!!!

jishuxu commented on pipelines/tasks/Kallisto.wdl:
```diff
@@ -0,0 +1,155 @@
+
+  runtime {
+    docker: "humancellatlas/kallisto:0.43.1"
+    cpu: 8
```
Should you add '--threads=8 '  in cmd? The default thread num is 1.

ambrosejcarr replied:
Added where possible; pseudobam does not support multiple threads, as discovered above.

jishuxu commented on pipelines/tasks/Kallisto.wdl:
```diff
@@ -0,0 +1,155 @@
+}
+
+
+task PseudoSingleEnd {
```
what's the difference  between 'pseudoSingleEnd' and 'QuantSingleEnd'?

ambrosejcarr replied:
Pseudo only runs pseudoalignment. Quant runs both Pseudoalignment and quantification. See: http://pachterlab.github.io/kallisto/manual.html

jishuxu commented on pipelines/tasks/Kallisto.wdl:
```diff
@@ -0,0 +1,155 @@
+}
+
+
+task PseudoSingleEndUMI {
```
what's the difference  between 'PseudoSingleEndUMI' and 'PseudoSingleEnd' ?

ambrosejcarr replied:
UMI counts number of **UMIs** in each compatibility group, the other counts the number of **reads**.

jishuxu commented on pipelines/tasks/Kallisto.wdl:
```diff
@@ -0,0 +1,155 @@
+    kallisto quant \
+      --index "${index}" \
+      --output-dir . \
+      --bootstrap-samples 100 \
```
I would add '--seed' to have fixed random number. it is just easier from testing point of view.

ambrosejcarr replied:
Default of 42 should produce a fixed seed, no? Is there a point in setting it to something else that I'm missing?

jishuxu replied:
copy paste from manual "--seed=INT                Seed for the bootstrap sampling (default: 42)"
http://pachterlab.github.io/kallisto/manual.html

jishuxu commented:
Overall it is great. the testing notebook is informative. 
Is it possible to check Kallisto quantification results, either by printing out several lines of count matrix or visualizing the count matrix.  I just want to make sure the Kallisto actually produces the results we expect.

ambrosejcarr commented:
Added a section to the notebook to visualize the results of each function (except index, whose functionality is implied by the successful exit of dependent methods). 

Added docstrings to each wdl method to make it clearer what they're designed to do.

jishuxu commented:
The updated notebook is exactly what I asked for. Thanks!

jishuxu commented:
PR looks good fine with me and it is okay to merge
